src/origin.js

```diff
@@ -1,16 +1,20 @@
-exports.match = function (incomingOrigin, origins) {
-  if (!incomingOrigin) {
-    return null
-  }
 
-  for (var i = 0; i < origins.length; i++) {
-    var origin = origins[i]
-    if ((origin instanceof RegExp && origin.test(incomingOrigin)) ||
-            (typeof origin === 'string' && origin === incomingOrigin) ||
-            (origin === '*')) {
-      return incomingOrigin
+exports.generic = function (list) {
+  return list.length === 1 && list[0] === '*'
+}
+
+exports.allowed = function (list, requestOrigin) {
+  function match (origin) {
+    if (origin.indexOf('*') !== -1) {
+      var regex = '^' + origin.replace('.', '\\.').replace('*', '.*') + '$'
```

Comment:
small performance comment, could we generate a list of these regex when the middleware is created rather then doing it on each request?